sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1209,67 +1215,137 @@ public void setTimestampPolicy(String timestampPolicy) {
       Coder<K> keyCoder = getKeyCoder(coderRegistry);
       Coder<V> valueCoder = getValueCoder(coderRegistry);
 
-      // The Read will be expanded into SDF transform when "beam_fn_api" is enabled.
-      if (!ExperimentalOptions.hasExperiment(input.getPipeline().getOptions(), "beam_fn_api")
-          || ExperimentalOptions.hasExperiment(
+      // For read from unbounded in a bounded manner, we actually are not going through Read or SDF.
+      if (ExperimentalOptions.hasExperiment(
               input.getPipeline().getOptions(), "beam_fn_api_use_deprecated_read")
           || getMaxNumRecords() < Long.MAX_VALUE
           || getMaxReadTime() != null) {
+        return input.apply(new ReadFromKafkaViaUnbounded<>(this, keyCoder, valueCoder));
+      }
+      return input.apply(new ReadFromKafkaViaSDF<>(this, keyCoder, valueCoder));
+    }
+
+    public static final PTransformOverride KAFKA_READ_OVERRIDE =
```

Comment:
Mark `@Internal` just to be clear that it is not for pipeline authors. Would be good to document why this exists and when to use it.

Perhaps it could be in runners-core-construction, but I actually want to merge that back into the core SDK so no need.

Reply:
runners-core-construction cannot depend on kafka-io because expansion-service introduces circular dependency, just like pubsub